baselines/flanders/flanders/strategy.py

```diff
@@ -0,0 +1,234 @@
+                initial_parameters = initial_parameters,
+                fit_metrics_aggregation_fn = fit_metrics_aggregation_fn,
+                evaluate_metrics_aggregation_fn = evaluate_metrics_aggregation_fn
+            )
+        self.dataset_name = dataset_name
+        self.warmup_rounds = warmup_rounds
+        self.to_keep = to_keep
+        self.window = window
+        self.maxiter = maxiter
+        self.alpha = alpha
+        self.beta = beta
+        self.params_indexes = None
+        self.malicious_selected = False
+        self.distance_function = distance_function
+
+    def configure_fit(
+        self, server_round: int, parameters: Parameters, client_manager: ClientManager
+    ) -> List[Tuple[ClientProxy, FitIns]]:
+        """Configure the next round of training."""
+        
+        # Sample clients
+        sample_size, min_num_clients = self.num_fit_clients(
+            client_manager.num_available()
+        )
+        
+        # Custom FitIns object for each client
+        fit_ins_list = [
+            FitIns(parameters, {} if not self.on_fit_config_fn else self.on_fit_config_fn(server_round)) 
+            for _ in range(sample_size)
+        ]
+
+        clients = client_manager.sample(
+            num_clients=sample_size, min_num_clients=min_num_clients
+        )
+
+        # Return client/config pairs
+        return [(client, fit) for client, fit in zip(clients, fit_ins_list)]
+        
+    def aggregate_fit(
+        self,
+        server_round: int,
+        results: List[Tuple[ClientProxy, FitRes]],
+        failures: List[Union[Tuple[ClientProxy, FitRes], BaseException]],
+        clients_state: Dict[int, bool],
+    ) -> Tuple[Optional[Parameters], Dict[str, Scalar]]:
+        """
+        Apply MAR forecasting to exclude malicious clients from FedAvg.
+
+        Parameters
+        ----------
+        server_round : int
+            Current server round.
+        results : List[Tuple[ClientProxy, FitRes]]
+            List of results from the clients.
+        failures : List[Union[Tuple[ClientProxy, FitRes], BaseException]]
+            List of failures from the clients.
+
+        Returns
+        -------
+        parameters_aggregated: Optional[Parameters]
+            Aggregated parameters.
+        metrics_aggregated: Dict[str, Scalar]
+            Aggregated metrics.
+        malicious_clients_idx: List[int]
+            List of malicious clients' cids (indexes).
+        """
+        good_clients_idx = []
+        malicious_clients_idx = []
+        if server_round > 1:
+            win = self.window
+            if server_round < self.window:
+                win = server_round
+            M = load_all_time_series(dir="clients_params", window=win)
+            M = np.transpose(M, (0, 2, 1))  # (clients, params, time)
+
+            M_hat = M[:,:,-1].copy()
+            pred_step = 1
+            print(f"aggregate_fit - Computing MAR on M {M.shape}")
```

Comment:
please replace all `print` statements in your baseline with `log(INFO, ...)` if you want them to be captured in the log. to do that you just need to import the below:
```python
from logging import INFO
from flwr.common.logger import log
```